stacker/dag/__init__.py

```diff
@@ -364,3 +365,92 @@ def size(self):
 
     def __len__(self):
         return len(self.graph)
+
+
+def walk(dag, walk_func):
+    return dag.walk(walk_func)
+
+
+def walk_threaded(dag, walk_func):
+    """ Walks each node of the graph, in parallel if it can.
+    The walk_func is only called when the nodes dependencies have been
+    satisfied
+    """
+
+    # First, we'll topologically sort all of the nodes, with nodes that
+    # have no dependencies first. We do this to ensure that we don't call
+    # .join on a thread that hasn't yet been started.
+    #
+    # TODO(ejholmes): An alternative would be to ensure that Thread.join
+    # blocks if the thread has not yet been started.
+    nodes = dag.topological_sort()
+    nodes.reverse()
+
+    # This maps a node name to a thread of execution.
+    threads = {}
+
+    # Blocks until all of the given nodes have completed execution (whether
+    # successfully, or errored). Returns True if all nodes returned True.
+    def wait_for(nodes):
+        return all(threads[node].join() for node in nodes)
+
+    # For each node in the graph, we're going to allocate a thread to
+    # execute. The thread will block executing walk_func, until all of the
+    # nodes dependencies have executed successfully.
+    #
+    # If any node fails for some reason (e.g. raising an exception), any
+    # downstream nodes will be cancelled.
+    for node in nodes:
+        def fn(n, deps):
+            if deps:
+                logger.debug(
+                    "%s waiting for %s to complete",
+                    n,
+                    ", ".join(deps))
+
+            # Wait for all dependencies to complete sucessfully.
+            if not wait_for(deps):
+                logger.debug("cancelling %s. "
+                             "Some dependencies "
+                             "were not satisfied", n)
+                return False
```

Comment:
Note to self. This should just continue to walk to the graph, instead of stopping. Conditional execution of the node is handled [further up the call stack](https://github.com/remind101/stacker/blob/044ca52e2453dbddcc54d4a72714962bb04842f6/stacker/plan.py#L349-L352), which is a better place to abort.